crates/accelerate/src/synthesis/linear/lnn.rs

```diff
@@ -0,0 +1,321 @@
+            if empty_layers > 1 {
+                // if nothing happened twice in a row, then finished.
+                done = true;
+            }
+        } else {
+            empty_layers = 0;
+        }
+        first_qubit = 1 - first_qubit;
+    }
+    cx_instructions_rows
+}
+
+///    Transform a north-west triangular matrix to identity in depth 3*n by Proposition 7.4 of [1]
+fn _north_west_to_identity(n: usize, mut mat: ArrayViewMut2<bool>) -> InstructionList {
+    // At start the labels are in reversed order
+    let mut label_arr: Vec<usize> = (0..n).rev().collect();
+    let mut first_qubit = 0;
+    let mut empty_layers = 0;
+    let mut done = false;
+    let mut cx_instructions_rows: InstructionList = Vec::new();
+    while !done {
+        let mut at_least_one_needed = false;
+        for i in (first_qubit..n - 1).step_by(2) {
+            // Exchange the labels if needed
+            if label_arr[i] > label_arr[i + 1] {
+                at_least_one_needed = true;
+                // If row i has "1" in column i+1, swap and remove the "1" (in depth 2)
+                // otherwise, only do a swap (in depth 3)
+                if !mat[[i, label_arr[i + 1]]] {
+                    // Adding this turns the operation to a SWAP
+                    _row_op_update_instructions(
+                        &mut cx_instructions_rows,
+                        mat.view_mut(),
+                        i + 1,
+                        i,
+                    );
+                }
+                _row_op_update_instructions(&mut cx_instructions_rows, mat.view_mut(), i, i + 1);
+                _row_op_update_instructions(&mut cx_instructions_rows, mat.view_mut(), i + 1, i);
+
+                (label_arr[i], label_arr[i + 1]) = (label_arr[i + 1], label_arr[i]);
+            }
+        }
+
+        if !at_least_one_needed {
+            empty_layers += 1;
+            if empty_layers > 1 {
+                // if nothing happened twice in a row, then finished.
+                done = true;
+            }
+        } else {
+            empty_layers = 0;
+        }
+        first_qubit = 1 - first_qubit;
+    }
+    cx_instructions_rows
+}
+
+///     Optimize CX circuit in depth bounded by 5n for LNN connectivity.
+///     The algorithm [1] has two steps:
+///     a) transform the original matrix to a north-west matrix (m2nw),
+///     b) transform the north-west matrix to identity (nw2id).
+///     
+///     A square n-by-n matrix A is called north-west if A[i][j]=0 for all i+j>=n
+///     For example, the following matrix is north-west:
+///     [[0, 1, 0, 1]
+///     [1, 1, 1, 0]
+///     [0, 1, 0, 0]
+///     [1, 0, 0, 0]]
+
+///     According to [1] the synthesis is done on the inverse matrix
+///     so the matrix mat is inverted at this step
+#[pyfunction]
+#[pyo3(signature = (mat))]
+pub fn optimize_cx_circ_depth_5n_line(
+    _py: Python,
+    mat: PyReadonlyArray2<bool>,
+) -> PyResult<(InstructionList, InstructionList)> {
```

Comment:
For the rust function to be entirely self-contained and for consistency with other synthesis functions, it would be nice to return `PyResult<CircuitData>`.

Reply:
Done in 4902815e1